File: leetcode/solutions/0746.MinCostClimbingStairs/minCostClimbingStairs.ts
// for loop version
export default function minCostClimbingStairs(cost: number[]): number {
  for (let i = 2; i < cost.length; i++) {
    cost[i] += Math.min(cost[i - 1], cost[i - 2]);
  }

  return Math.min(cost[cost.length - 1], cost[cost.length - 2]);
}

// dp version
function minCostClimbingStairs2(cost: number[]): number {
  let dp = [0, cost[0], cost[1]];

  for (let i = 2; i < cost.length; i++) {
    dp = [dp[1], dp[2], Math.min(dp[1], dp[2]) + cost[i]];
  }

  return Math.min(dp[1], dp[2]);
}

File: leetcode/solutions/0746.MinCostClimbingStairs/minCostClimbingStairs.go
package solution

func minCostClimbingStairs(cost []int) int {
	if len(cost) <= 1 {
		return 0
	}

	dp := [...]int{cost[0], cost[1]}

	for i := 2; i < len(cost); i++ {
		dp[0], dp[1] = dp[1], min(dp[0], dp[1])+cost[i]
	}

	return min(dp[0], dp[1])
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}
